xarray/core/treenode.py

```diff
@@ -44,6 +45,68 @@ def __init__(self, *pathsegments):
 Tree = TypeVar("Tree", bound="TreeNode")
 
 
+class Children(Mapping[str, Tree], Generic[Tree]):
+    """
+    Dictionary-like container for the immediate children of a single DataTree node.
```

Comment:
I'm not sure whether this class should allow path-like access to grandchildren or only allow access to immediate children.

Reply:
Same question also for setting

Reply:
On second thoughts I think this class should only allow getting/setting immediate children.

Reply:
> On second thoughts I think this class should only allow getting/setting immediate children (because it's basically a glorified `dict`).

Agreed!

Reply:
This relates to https://github.com/pydata/xarray/issues/9485 though - why exactly do we want to support full paths for `.coords` but only local access for `.children`? We could do whatever, but what's the rationale for them being different?

Reply:
That's a fair question!

From a user perspective, I would say we should definitely be consistent. And supporting full paths is certainly convenient, and doesn't add much overhead.

Reply:
So you're basically thinking of `.coords`/`.children` as meaning "access to all coords/children defined anywhere on the subtree". I originally thought of them as "access to just coords/children defined on this node".

FYI path-like access like this means that `list(dt.children)` will return a subset of what can be accessed via `.__getitem__`, (and `__contains__` could go either way, see https://github.com/pydata/xarray/issues/9354)).

Reply:
One compromise option of sorts (similar to what is done for Dataset for derived variables like `time.year`, or for coordinates in `Dataset.__getitem__`) is to only list immediate children in `__iter__` and `__contains__` but to support accessing them in `__getitem__`.

Reply:
There is a detailed discussion about this on the old repo (https://github.com/xarray-contrib/datatree/issues/240#issuecomment-1527735223). There the conclusion was that `__contains__` should support paths if `__getitem__` does, but it's okay for `.keys()`/`.values()`/`.items()` to be local only.

> So you're basically thinking of .coords/.children as meaning "access to all coords/children defined anywhere on the subtree".

To play devil's advocate a bit: If `.coords` can access the whole tree then why not also `.variables`? I think they should all be consistent.

Reply:
> There is a detailed discussion about this on the old repo ([xarray-contrib/datatree#240 (comment)](https://github.com/xarray-contrib/datatree/issues/240#issuecomment-1527735223)). There the conclusion was that `__contains__` should support paths if `__getitem__` does, but it's okay for `.keys()`/`.values()`/`.items()` to be local only.

Yes, that seems to be how Dataset works for coordinates.

> To play devil's advocate a bit: If .coords can access the whole tree then why not also .variables? I think they should all be consistent.

Sure, we could do that. Hopefully it's all very straightforwardly reusing the same machinery?

Reply:
> Yes, that seems to be how Dataset works for coordinates.

I'm not sure what you mean. You're talking about virtual vs non-virtual coordinates on `Dataset`? That they are all in `__contains__` and `__getitem__`, but only a subset (the non-virtual ones) are listed by `.keys()`?

> > If `.coords` can access the whole tree then why not also `.variables`?
>
> Sure, we could do that. 

I actually meant that as a example of counter-intuitive behaviour 😅 . But maybe it's fine too? That would mean we have really done a 180 degree turn in this issue... It would imply changing the current behaviour of `dt.data_vars` too.

> Hopefully it's all very straightforwardly reusing the same machinery?

It's not right now (local vs path-supporting have different codepaths), but it could be made to: everything could go through some version of using `TreeNode._set_item`

https://github.com/pydata/xarray/blob/8bd456cdb5a873684ed18e77fe4f6983738084db/xarray/core/treenode.py#L550

or [`._walk_to`](https://github.com/pydata/xarray/issues/9485#issuecomment-2345203393) and we could add a parameter like `traverse_paths` to generalize it for both cases. Generalizing `.update` to support paths might be the approach that is easiest to integrate with the `Coordinates` base class.

Also one final thing: We are talking here about supporting paths to descendant nodes only? Or upwards via `dt['../../path/to/node/']` too?